### ISSUE TITLE
Server side pagination with search

### ISSUE THREAD

**Issue by gjakab** [cult-of-coders/grapher#251]
I have a webpage where I'm listing some data from the database. There's also a button that the users can press to load more stuff from the database. I'm trying to limit my query so not all the data get's loaded to the client at once. The users can also search in the data.

My Query:
```javascript
export default createQuery({
  movies: {
    $filter({filters, options, params}) {
      filters.title = { $regex: new RegExp(`.*${params.title}.*`, "i") };
    },
    $paginate: true,
    title: 1
  }
}); 
```
Client:
```javascript
export default withQuery((props) => {
  return query.clone({
    title: props.title,
    limit: props.limit
  });
}, {reactive: true})(Movies)
```
My problem is that the search only occurs in the limited documents and not in every document in the movies collection.
Is there any way to fix this?

**Comment by neoromantic** [comment]
Also, paging doesn't work with reactive queries.

**Comment by theodorDiaconu** [comment]
@gjakab how are you increasing this limit ? Could you show me a quick repro ?

**Comment by theodorDiaconu** [comment]
@neoromantic how so? can you provide more details, maybe a reproduction ?

**Comment by theodorDiaconu** [comment]
@gjakab @neoromantic any details please ? I'm gonna close the issue because I have no clue what the issue is.

**Comment by Twisterking** [comment]
I think I am having the same problem. For some reason, using a regex in a grapher query **does not always return all the results**. 
My query `$filter` looks a bit like this:
```js
  $filter({filters, options, params}) {
    if(params.filter) {
      Object.keys(params.filter).forEach(key => {
        if(key == 'searchTerm') {
          let searchTerm = params.filter[key];
          let words = searchTerm.trim().split(/[ \-\:]+/);
          let regex = buildRegExp(words, false);
          filters.$or = [
            { item_name: regex },
            { item_ref: regex }
          ];
        } else {
          let param = params.filter[key];
          if(param) filters[key] = param;
        }
      });
    }
  }
```
with
```js
function buildRegExp(words, startsWith) {
	var exps = _.map(words, function(word, key) {
		if(startsWith && key == 0)
			return "(?=^" + word + ")";
		else
			return "(?=.*" + word + ")";
	});
	var fullExp = exps.join('') + ".+";
	return new RegExp(fullExp, "i");
}
```
On the client side I **always clone the query with current params with `{ reactive: true }`** - but still, the search is very inconsistent (and way slower than `Meteor.subscribe`ing to a publication which just returns the results and uses Mongo indices.

For example: Copy-pasting the whole text "Gurke" into my search input properly returns all the items containing "Gurke" in the `item_name` or `item_ref`. Deleting the whole "Gurke" inside the input and copy-pasting "Tomate" also works properly. If I now use the backspace key, delete everything and start typing "G", then "Gu", then "Gur" (...) while using a 200ms debounce, the **results are not the same as just removing everything again and copy-pasting "Gurke" again**!

**Comment by theodorDiaconu** [comment]
@Twisterking do you have a namedQuery or global query ?

**Comment by Twisterking** [comment]
It is a global query!

**Comment by theodorDiaconu** [comment]
@Twisterking sorry to hear that, maybe you forgot to expose them? Global queries are dangerous to secure.

And yes, I think it's related to the fact that the regexp gets cloned and bits of it may be missing. Please go with named queries.

**Comment by Twisterking** [comment]
I just tried my best but can't get it to work! :(

Can you provide an example on how to use a named query with https://github.com/cult-of-coders/grapher-react? For some reason I always end up with some random error.

I do not understand how to really write named queries (do they have to live in separate files inside a `queries` folder?) and also how to reuse them on the (React) client side.

**Comment by Twisterking** [comment]
weeeeellll nevermind - I got it to work! 👍 
One of those cases - should have properly read the docs! 😆

**Comment by martineboh** [comment]
@Twisterking Please if you can, share how you got it working correctly, thanks.

**Comment by Twisterking** [comment]
Okay so what I did wrong is the following:
The named queries (as far as I can tell) **have to** live in the folder `/imports/db/<collection>/queries/<queryname>`. From there you have to `export default` the named query, import it again in your `/imports/db/<collection>/expose.js` file and `<queryname>.expose()` them.
Important:
You have to also create a file, mine is called `/imports/db/queries.js` where you import ALL your named queries. This `/imports/db/queries.js` file has to be imported in your `/imports/startup/client.js` file (or somewhere on the client) too. Without that (that was my big mistake btw) the client has no knowledge of the named queries.
Hope this helps!

**Comment by theodorDiaconu** [comment]
@Twisterking usually what we do is we have exposures.js which includes all exposures. And namedQueries should be used as modules (import them directly)

**Comment by Twisterking** [comment]
Running into problems again! :(
As far as I can tell I did not change anything really and suddenly my paginated queries always return the same results. They behave very very weirdly.
Named query, `import`ed as you said, `clone()`d it, I tried pretty much everything but it does not really paginate as it should.

**Comment by theodorDiaconu** [comment]
Named Query don't support filters and options by default like global queries do, you have to implement your own function for that, inside $filter(), please do a rundown of the documentation, top to bottom, I feel that you have skipped it. 

Using $filter() function you can now use `filters` and `options` by applying them from params, however keep in mind that with Meteor, multiple subscriptions result in merged dataset, so if you don't have the client-side query-ing done right, you will endup with more results in one place, this is why we need Grapher to do a client-side filtering as well.

In order to do client-side filtering it needs to know the filters. Sometimes this exposes the data model, for example, you filter by the user not being an admin, if you want the client to do that, it needs to be exposed somehow to extract it to the collection's dataset currently in the client.

For now, you can do some manual work and use sift library to perform your client-side filtering of Grapher's returned data. 
https://www.npmjs.com/package/sift to filter again received data from Grapher.

**Comment by paulincai** [comment]
Thanks for your time. I am past filters and all that, all implemented via params, I got that part, it's done.

My problem is simply that I don't get which part does the publication. Just to mentioned again simply, I need most recent documents sorted the other way around which can be easily done by filed sorting pub -1 and sub +1. Not sure how to combine the 2 in Grapher.

I moved all my other reactive (5%) and non reactive (the rest) queries to Grapher, it's not really like I am taking the first loot at it but I'm just not getting this particular situation.

**Comment by Twisterking** [comment]
Thanks a bunch for your fast reply as always @theodorDiaconu !
I am using `$filter()` already and it works properly - I think.

If I understood you correctly, you are saying that due to Meteor's subscription handling behind the scenes, I have to use `$postFilter` or `$postFilters (sift)` to filter the data on the client again? Is this correct?

**Comment by theodorDiaconu** [comment]
@Twisterking I perfectly understand your problem, but this is a known and old issue from Meteor PubSub. Given 2 publications with collision on documents, you may encounter trouble trying to filter them.

**Comment by bhunjadi** [comment]
@Twisterking @theodorDiaconu 

I think I had something similar for an issue and believe I solved it by using scoped queries. It is an issue of pub-sub, but it seems to me the problem can be overcome with scoped queries.

I tried that idea in my own fork of grapher and it worked as expected. I used library from mitar, [meteor-subscription-scope](https://github.com/peerlibrary/meteor-subscription-scope), and made some adaptation for grapher.

Example of usage:
```
const query = createQuery('...', {
    $filters: {...},
    ...,
}, {
    scoped: true,
});
```

Notice the scoped option for a query.

@theodorDiaconu 
If you are interested, I can make PR.
This is the [commit](https://github.com/bhunjadi/grapher/commit/a673f16d1fd2569496e291fe1425a9082ffb0705) where I implemented that feature.
